### PR TITLE
C-Chain Transaction Serialization

### PR DESCRIPTION
**Pull request by cgcardona**
This PR adds transaction serialization for the EVM/C-Chain. It follows a similar architecture to the other 2 virtual machines but due to slight differences between the transactions on the PlatformVM/AVM and the EVM we are unable to use the existing `BaseTx` class hierarchy and needed to partially recreate it as `EVMBaseTx`.

The primary differences between the PlatformVM/AVM and EVM transaction formats are as follows:

PlatformVM/AVM `BaseTx`s have this format:

* `CodecID`
* `TypeID`
* `NetworkID`
* `BlockchainID`
* `Outputs`
* `Inputs`
* `Memo`

EVM transactions have this format:

* `CodecID`
* `TypeID`
* `NetworkID`
* `BlockchainID`
* `SourceChain/DestinationChain`
* `Inputs`
  * `EVMInputs`
  * `TransferableInputs`
* `Outputs`
  * `EVMOutputs`
  * `TransferableOutputs`

It's possible we could have altered the PlatformVM/AVM `BaseTx` with conditionals to potentially work on all 3 VMs but I decided against that because it would have involved making major changes to the PlatformVM and AVM and that seemed to introduce too much risk. It seemed like a safer move to duplicate some of the `BaseTx` code and leave the PlatformVM/AVM code untouched.

### Changes Include

[EVMAPI](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/api.ts) is for JSON RPC calls to AvalancheGo. It includes: `export`, `exportAVAX`, `getUTXOs`, `import`, `importAVAX`, `importKey`, `issueTx`, and `exportKey`.

[EVM Base Tx](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/basetx.ts) is a base class which is extended by `ImportTx` and `ExportTx`. It contains `TypeID`, `NetworkID` and `BlockchainID`

[EVM Constants](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/constants.ts)

[EVM Credentials](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/credentials.ts). The only type of EVM Credential is `SECPCredential`

[EVM Export Tx](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/exporttx.ts). This extends `EVMBaseTx` and additionally has `DestinationChain`, `Inputs` which is an array of `EVMInputs` and `exportedOutputs` which is an array of `TransferableOutputs`.

[EVM Import Tx](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/importtx.ts). This extends `EVMBaseTx` and additionally has `SourceChain`, `ImportedInputs` which is an array of `TransferableInputs` and `Outs` which is an array of `EVMOutputs`.

[EVM Keychain](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/keychain.ts) This is copied over from the PlatformVM and AVM Keychains.

[EVM Inputs](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/inputs.ts)
which contain `Address`, `Amount`, `AssetID` and `Nonce`

[EVM Outputs](https://github.com/ava-labs/avalanchejs/blob/evm-serialization/src/apis/evm/outputs.ts)
which contain `Address`, `Amount`, `AssetID`

### References

[C-Chain Transaction Serialization](https://docs.avax.network/build/references/coreth-atomic-transaction-serialization)

### Testing Scripts

* [ImportTx-avax](https://gist.github.com/cgcardona/1b49a998bff47857707ce491102795bb)
* [ImportTx-ant](https://gist.github.com/cgcardona/7888889e07367897b2770a815617601d)
* [ExportTx-avax](https://gist.github.com/cgcardona/147f47c88e92de061599400eb06d7e7f)
* [ExportTx-ant](https://gist.github.com/cgcardona/ae1e15dfde9abe20f3b29527094941c9)
* [cchain.buildImportTx.ts](https://gist.github.com/cgcardona/88c52c30145a0c7d79165266d66d2f0f)
* [cchain.buildExportTx-avax.ts](https://gist.github.com/cgcardona/4824c4632e52099e1ff327b3ba16e138)
* [cchain.buildExportTx-ant.ts](https://gist.github.com/cgcardona/2b79fccaa1b24f2a572902f688acbebb)

### Testing Instructions

* [Importing and Exporting AVAX from the C-Chain using AvalancheJS](https://gist.github.com/cgcardona/423be4a3761d1e97dd7a2deb31340e45)
* [Importing and Exporting AVAX and ANTs from the C-Chain using AvalancheJS](https://gist.github.com/cgcardona/4de6ff0877fc0162a8150ac85581772a)

### Remaining Steps

* Flesh out automated testsuite
* Manual testing
* `serialize` and `deserialize` for all classes which extend `Serializable`
* Documentation
* Sanity-check coding
* Address security audit from Arnold

### CODE REVIEW

File: src/apis/evm/tx.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    this.credentials = [];
+    for(let i = 0; i < fields["credentials"].length; i++){
+      const cred:Credential = SelectCredentialClass(fields["credentials"][i]["_typeID"]);
+      cred.deserialize(fields["credentials"][i], encoding);
+      this.credentials.push(cred);
+    }
+  }
+
+  /**
+   * Takes a {@link https://github.com/feross/buffer|Buffer} containing an [[Tx]], parses it, populates the class, and returns the length of the Tx in bytes.
```
Comment (dasconnor) [inline src/apis/evm/tx.ts:104]: nit: line length

File: src/common/evmtx.ts [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+   * Returns the NetworkID as a number
+   */
+  getNetworkID = ():number => this.networkid.readUInt32BE(0);
+
+  /**
+   * Returns the Buffer representation of the BlockchainID
+   */
+  getBlockchainID = ():Buffer => this.blockchainid;
+
+//   /**
```
Comment (dasconnor) [inline src/common/evmtx.ts:62]: Can we clean this up?

File: src/common/evmtx.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    codecBuf.writeUInt16BE(this.codecid, 0);
+    return codecBuf;
+  } 
+
+  /**
+   * Returns the inputTotal as a BN 
+   */
+  getInputTotal = (assetID:Buffer):BN=> {
+    const ins:Array<StandardTransferableInput> = [];
+    // const ins:Array<StandardTransferableInput> = this.getTransaction().getIns();
```
Comment (dasconnor) [inline src/common/evmtx.ts:161]: delete here and in next method

File: src/apis/evm/utxos.ts [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+         const evmOutput: EVMOutput = new EVMOutput(
+           toAddresses[0],
+           amt,
+           assetID
+         );
+         outs.push(evmOutput);
+       }
+     });
+     
+     // TODO - review commented out code to confirm that we don't need it.
```
Comment (dasconnor) [inline src/apis/evm/utxos.ts:325]: another todo

**Review by dasconnor — COMMENTED** [review]
Couple of small code quality suggestions. There are a number of TODOs in the code. Can you either resolve those or create Jira tickets for them? Great work!

File: src/apis/evm/importtx.ts [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  /**
+     * Takes the bytes of an [[UnsignedTx]] and returns an array of [[Credential]]s
+     *
+     * @param msg A Buffer for the [[UnsignedTx]]
+     * @param kc An [[KeyChain]] used in signing
+     *
+     * @returns An array of [[Credential]]s
+     */
+  sign(msg: Buffer, kc: KeyChain): Credential[] {
+    // const sigs:Array<Credential> = super.sign(msg, kc);
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/importtx.ts]: Line deleted

File: src/common/evmtx.ts (hunk 3 above)
Comment (cgcardona) [reply src/common/evmtx.ts]: Line deleted

File: src/apis/evm/api.ts [hunk 6]
```diff
@@ -21,8 +40,169 @@ const bintools:BinTools = BinTools.getInstance();
...
+   * Refresh blockchainID, and if a blockchainID is passed in, use that.
+   *
+   * @param Optional. BlockchainID to assign, if none, uses the default based on networkID.
+   *
+   * @returns The blockchainID
+   */
+  refreshBlockchainID = (blockchainID: string = undefined): boolean => {
+    const netID: number = this.core.getNetworkID();
+    if (typeof blockchainID === 'undefined' && typeof Defaults.network[netID] !== "undefined") {
+      // TODO - confirm `blockchainID` is working instead of `blockchainID`
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/api.ts]: That `TODO` was awkwardly phrased. What I meant to say was 

> confirm if `blockchainID` is passed in and isn't passed in.

So now I have two `if` statements to check for that. This is resolved.

File: src/apis/evm/api.ts [hunk 7]
```diff
@@ -60,16 +242,63 @@ export class EVMAPI extends JRPCAPI {
...
-    return this.callMethod('avax.exportAVAX', params).then((response:RequestResponseData) => response.data.result.txID);
+    this.setBaseURL("/ext/bc/C/avax");
+    return this.callMethod('avax.exportAVAX', params).then((response: RequestResponseData) => response.data.result.txID);
   };
 
+  /**
+   * Retrieves the UTXOs related to the addresses provided from the node's `getUTXOs` method.
+   *
+   * @param addresses An array of addresses as cb58 strings or addresses as {@link https://github.com/feross/buffer|Buffer}s
+   * @param sourceChain A string for the chain to look for the UTXO's. Default is to use this chain, but if exported UTXOs exist from other chains, this can used to pull them instead.
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/api.ts]: Resolved

File: src/apis/evm/exporttx.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+   */
+  toBuffer(): Buffer {
+    if(typeof this.destinationChain === "undefined") {
+      throw new Error("ExportTx.toBuffer -- this.destinationChain is undefined");
+    }
+    this.numInputs.writeUInt32BE(this.inputs.length, 0);
+    this.numExportedOutputs.writeUInt32BE(this.exportedOutputs.length, 0);
+    let barr: Buffer[] = [super.toBuffer(), this.destinationChain, this.numInputs];
+    let bsize: number = super.toBuffer().length + this.destinationChain.length + this.numInputs.length;
+    // this.importIns = this.importIns.sort(TransferableInput.comparator());
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/exporttx.ts:79]: Good catch. Resolved.

File: src/apis/evm/importtx.ts [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  /**
+   * Returns a {@link https://github.com/feross/buffer|Buffer} for the source chainid.
+   */
+  getSourceChain = (): Buffer => {
+    return this.sourceChain;
+  }
+
+  /**
+     * Takes a {@link https://github.com/feross/buffer|Buffer} containing an [[ImportTx]], parses it, populates the class, and returns the length of the [[ImportTx]] in bytes.
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/importtx.ts:71]: Resolved

[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/importtx.ts:71]: We should be able to set this and automate w/ a linter. I'll add it as a task.

File: src/apis/evm/keychain.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+  /**
+    * Returns the chainID associated with this key.
+    *
+    * @returns The [[KeyPair]]'s chainID
+    */
+  getChainID = (): string => this.chainid;
+
+  /**
+   * Sets the the chainID associated with this key.
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/keychain.ts]: Resolved

File: src/apis/evm/outputs.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+  /**
+   * Returns the address of the input as {@link https://github.com/feross/buffer|Buffer}
+   */
+  getAddress = (): Buffer => this.address;
+
+  /**
+   * Returns the address as a bech32 encoded string.
+   */
+  // TODO - Get getAddressString to work. Why is `getPreferredHRP(networkID)` failing?
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/outputs.ts]: Addresses are hex encoded in `EVMOutput`. We don't need to return the address as bech32 so we don't need to debug `getPreferredHRP`. This is now resolved.

File: src/apis/evm/tx.ts (hunk 1 above)
Comment (cgcardona) [reply src/apis/evm/tx.ts]: Resolved

File: src/apis/evm/tx.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * Takes a buffer representing the output and returns the proper [[BaseTx]] instance.
+ *
+ * @param txtype The id of the transaction type
+ *
+ * @returns An instance of an [[BaseTx]]-extended class.
+ */
+export const SelectTxClass = (txtype:number, ...args:Array<any>):EVMBaseTx => {
+  // if (txtype === EVMConstants.BASETX) {
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/tx.ts:34]: Resolved

File: src/common/evmtx.ts (hunk 2 above)
Comment (cgcardona) [reply src/common/evmtx.ts:62]: Resolved

File: src/apis/evm/basetx.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+   * Takes the bytes of an [[UnsignedTx]] and returns an array of [[Credential]]s
+   *
+   * @param msg A Buffer for the [[UnsignedTx]]
+   * @param kc An [[KeyChain]] used in signing
+   *
+   * @returns An array of [[Credential]]s
+   */
+  sign(msg: Buffer, kc: KeyChain): Credential[] {
+    const sigs: Credential[] = [];
+    // for (let i = 0; i < this.ins.length; i++) {
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/basetx.ts]: Now calling `super.sign` for both `ImportTx` and `ExportTx` to pass back `Credential` array. Resolved.

File: src/apis/evm/api.ts [hunk 14]
```diff
@@ -45,6 +225,8 @@ export class EVMAPI extends JRPCAPI {
       password,
       assetID
     };
+    // TODO - confirm calling `this.setBaseURL` isn't causing any unwanted side effects
```
[the remark this answers is not in the document]
Comment (cgcardona) [reply src/apis/evm/api.ts]: Only calling `setBaseURL` in one place now. This is resolved.

File: src/apis/evm/utxos.ts (hunk 4 above)
Comment (cgcardona) [reply src/apis/evm/utxos.ts]: Resolved.

**Pull request merged**